nvme_gw_persistence.py

```diff
@@ -0,0 +1,404 @@
+        """Reads single key from persistent config and returns its value."""
+
+        with rados.ReadOpCtx() as read_op:
+            iter, ret = self.ioctx.get_omap_vals_by_keys(read_op, (key,))
+            if ret != 0:
+                raise Exception("Omap read operation failed.")
+            self.ioctx.operate_read_op(read_op, self.omap_name)
+            value_list = list(dict(iter).values())
+            if len(value_list) == 1:
+                val = str(value_list[0], "utf-8")
+                self.logger.debug(f"Read key: {key} -> {val}")
+                return val
+        return None
+
+    def _read_all(self) -> Dict[str, str]:
+        """Reads persistent config and returns dict of all keys and values."""
+
+        with rados.ReadOpCtx() as read_op:
+            iter, ret = self.ioctx.get_omap_vals(read_op, "", "", -1)
+            if ret != 0:
+                raise Exception("Omap read operation failed.")
+            self.ioctx.operate_read_op(read_op, self.omap_name)
+            omap_dict = dict(iter)
+            self.logger.debug(f"Omap Persistent Config:\n{omap_dict}")
+        return omap_dict
+
+    def delete_config(self):
+        """Deletes OMAP object."""
+
+        try:
+            self.ioctx.remove_object(self.omap_name)
+            self.logger.info(f"Object {self.omap_name} deleted.")
+        except rados.ObjectNotFound:
+            self.logger.info(f"Object {self.omap_name} not found.")
+
+    def _clean_args(self, data: str) -> Dict[str, str]:
+        """Transforms configuration details from gRPC request format to a 
+        dictionary. Requires data string in key:value form."""
+
+        param_dict = {}
+        for arg in data.split('\n'):
+            if arg != "":
+                key, value = arg.split(':', 1)
+                param_dict[key.strip(' \"')] = value.strip(' \"')
+        return param_dict
+
+    def restore(self, callbacks):
+        """Restores gateway config to persistent config specifications."""
+
+        omap_version = self._read_key(self.OMAP_VERSION_KEY)
+        if omap_version == "1":
+            self.logger.info("This omap was just created. Nothing to restore")
+        else:
+            omap_dict = self._read_all()
+            self._restore_bdevs(omap_dict, callbacks[self.BDEV_PREFIX])
+            self._restore_subsystems(omap_dict,
+                                     callbacks[self.SUBSYSTEM_PREFIX])
+            self._restore_namespaces(omap_dict,
+                                     callbacks[self.NAMESPACE_PREFIX])
+            self._restore_hosts(omap_dict, callbacks[self.HOST_PREFIX])
+            self._restore_allow_all(omap_dict, callbacks[self.ALLOW_ALL_PREFIX])
+            self._restore_transports(omap_dict,
+                                     callbacks[self.TRANSPORT_PREFIX])
+            self._restore_listeners(omap_dict, callbacks[self.LISTENER_PREFIX])
+            self.version = omap_version
+            self.logger.info("Restore complete.")
```

Comment:
I would prefer if the gateway would keep in memory all dependencies between load/used entities. E.g. bdevs are created/restored the list of successful bdevs is kept, then when a namespace is created/restored the `bdevs` list is checked and the command is proceed only if bdev is in the list.

This might be improved in a further PR though.

Reply:
In this example we are logging bdev creation errors instead of aborting? I think your solution is addressing the issue caused when there is some problem with the omap specification. Please correct me if I'm misinterpreting your intention here. In that situation, we would need to fix the omap itself so it can be accurate for all gateways within a group. I think this is something to handle in a further PR, but further discussion will be helpful.

Reply:
I created issue #11 to continue the conversation on logging errors instead of aborting and keeping an internal map.